Compiler bug on compile-time with returned type

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 21.04
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
CC version: cc (Ubuntu 10.3.0-1ubuntu1) 10.3.0

getwd: /var/www/vproject/test/testblog
vmodules: /home/tanguy/.vmodules
vroot: /var/www/vproject/v
vexe: /var/www/vproject/v/v
vexe mtime: 2022-01-08 07:12:20
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 c1711b8.ac7e79f

Git version: git version 2.30.2
Git vroot status: weekly.2022.01-2-gac7e79ff
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 3654d6a8

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg /var/www/vproject/test/pattern_framwork/0sample.v`
```v
module main

struct S1 {}

fn  main() {
	s1 := S1{}
	$for method in S1.methods {
		println(s1.$method('ll'))
	}
}

fn (t S1) m(s string) int {
	return 7
}

```

**What did you expect to see?**

A formatted error


**What did you see instead?**
```
==================
/tmp/v_1000/0sample.13252888507795522193.tmp.c:10950: error: cannot convert 'int' to 'struct string'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -869,6 +869,15 @@ fn (mut g Gen) method_call(node ast.CallExpr) {
 					return
 				}
 			}
+		} else if node.left is ast.ComptimeCall {
+			if node.left.method_name == 'method' {
+				sym := g.table.sym(g.unwrap_generic(node.left.left_type))
+				if m := sym.find_method(g.comptime_for_method) {
+					rec_type = m.return_type
+					g.gen_expr_to_string(node.left, rec_type)
+					return
+				}
+			}
 		} else if node.left is ast.Ident {
 			if node.left.obj is ast.Var {
 				if g.comptime_var_type_map.len > 0 {
@@ -1217,6 +1226,13 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 						if expr.field_expr.expr is ast.Ident {
 							key_str := '${expr.field_expr.expr.name}.typ'
 							typ = g.comptime_var_type_map[key_str] or { typ }
+						}
+					}
+				} else if expr is ast.ComptimeCall {
+					if expr.method_name == 'method' {
+						sym := g.table.sym(g.unwrap_generic(expr.left_type))
+						if m := sym.find_method(g.comptime_for_method) {
+							typ = m.return_type
 						}
 					}
 				} else if expr is ast.Ident {

--- a/vlib/v/tests/comptime_for_method_call_in_print_call_test.v
+++ b/vlib/v/tests/comptime_for_method_call_in_print_call_test.v
@@ -1,0 +1,34 @@
+module main
+
+struct S1 {}
+
+fn test_comptime_for_method_call_str() {
+	s1 := S1{}
+	mut rets := []string{}
+
+	$for method in S1.methods {
+		println(s1.$method('ll'))
+		rets << s1.$method('ll').str()
+	}
+	assert rets.len == 4
+	assert rets[0] == '7'
+	assert rets[1] == '7.7'
+	assert rets[2] == 'foo'
+	assert rets[3] == 'true'
+}
+
+fn (t S1) m1(s string) int {
+	return 7
+}
+
+fn (t S1) m2(s string) f64 {
+	return 7.7
+}
+
+fn (t S1) m3(s string) string {
+	return 'foo'
+}
+
+fn (t S1) m4(s string) bool {
+	return true
+}

